src/lib/uTop.ml

```diff
@@ -32,8 +38,29 @@ let version = "%%VERSION%%"
    | History                                                         |
    +-----------------------------------------------------------------+ *)
 
+(** create a directory but doesn't raise an exception if the directory already exist *)
+let mkdir_safe dir perm =
+    try Unix.mkdir dir perm with Unix.Unix_error (Unix.EEXIST, _, _) -> ()
+
+(** create a directory, and create parent if doesn't exist *)
+let mkdir dir perm =
+    let rec p_mkdir dir =
+        let p_name = Filename.dirname dir in
+        if p_name <> "/" && p_name <> "."
+    then p_mkdir p_name;
+    mkdir_safe dir perm in
+    p_mkdir dir
+
 let history = LTerm_history.create []
-let history_file_name = ref (Some (Filename.concat LTerm_resources.home ".utop-history"))
+
+let history_file_name = ref (Option.map (fun cache_dir -> Filename.concat cache_dir "history") M.cache_dir)
+
+let () = Option.iter (fun cache_file ->
+    let dir = Filename.dirname cache_file in
+    if not (Sys.file_exists dir) then begin
```

Comment:
The `Sys.file_exists` is not required because `mkdir` does not fail if the directory exists.